rfc/graph-feature-flags.md

```diff
@@ -0,0 +1,242 @@
+
+Feature flags allow you to change the behavior of your API at runtime.
+Depending on the request context, e.g. through a Header, JWT Claim, or a cookie, features can be enabled or disabled.
+This allows you to create personalized experiences for different users without having to deploy multiple Routers or APIs.
+
+### Staging Environments
+
+Feature flags allow you to test new features in a staging environment before rolling them out to production,
+but more importantly, developers can test new features in a "real" environment without getting in the way of other developers.
+For larger environments, it's not possible to spin up the whole infrastructure on a developer's machine.
+If you want to test a new feature end-to-end, you need to have a complete staging environment that is as close to production as possible.
+This means that you'll very likely have one or more frontend applications, and many Subgraphs that are part of the Federated API.
+To be able to test a new feature, you need the entire stack to be running, so it's unlikely that you can run the whole stack for each feature in isolation.
+As a consequence, you need a shared staging environment where developers can test their features without affecting other developers,
+hence the need for feature flags so that individual developers can test their changes in as much isolation as possible.
+
+In a staging environment, you can enable a feature flag by setting a specific cookie or header in your client.
+
+## Summarizing the Problem
+
+If we condense the four main reasons, we can summarize the problem as two main points:
+
+1. **Feature Rollout**: We need a way to gradually roll out new features to our users
+2. **Experimental Features**: We need a way to test new features without affecting the "main" Graph
+
+As the two problems are slightly different,
+let's discuss them one by one,
+even though the solution to both problems is very similar: Feature Flags.
+
+## Gradual Feature Rollout: The Solution
+
+### Publishing Subgraphs with a Feature Flag & Enabling it Dynamically
+
+Here's the default cmd to publish a new Subgraph Schema:
+
+```bash
+npx wgc subgraph publish products --schema ../demo/subgraphs/products/products.graphql --routing-url http://localhost:4001/graphql --label=team=A
+```
+
+Let's say, we'd like to use the `@override` directive that overrides a field from another Subgraph,
+and we want to gradually roll out this new feature to our users.
+
+We can add a `--feature-flag` flag to the `subgraph publish` command:
+
+```bash
+npx wgc subgraph publish products --schema ../demo/subgraphs/products/products.graphql --routing-url http://localhost:4001/v2/graphql --label=team=A --feature-flag=v2
+```
+
+This will publish the new Subgraph Schema under the feature flag `products-v2`.
+Existing clients will continue to use the previous version of the Schema.
+
+To enable the feature flag for a specific request, you can set a cookie or header in your client.
```

Comment:
There could be a nice integration in the Cosmo Playground where you toggle the respective feature flags then.

Reply:
I just had this very idea.
How about using an alternative way to represent feature flags?
We could send them in the client extension:

```json
{
  "query": "{ products { id name } }",
  "extensions": {
    "featureFlags": [
      "products.experimental"
    ]
  }
}
```

Reply:
We could also make FF introspect-able, so that the Playground can fetch them to show you FF toggles:

```json
{
  "query": "{ __featureFlags { name description } }",
}
```

This will return all feature flags that are currently enabled in the staging environment.

```json
{
  "data": {
    "__featureFlags": [
      {
        "name": "products.experimental",
        "description": "Enables the experimental feature for the products Subgraph",
      }
    ]
  }
}
```